core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts

```diff
@@ -145,12 +217,92 @@ oppia.directive('questionPlayer', [
             return ctrl.totalQuestions;
           };
 
+          var calculateScorePerSkill = function(questionSkillData,
+              questionScores) {
+            $scope.scorePerSkill = [];
+            $scope.totalScore = 0.0;
+            var minScore = Number.MAX_VALUE;
+            var totalScore = 0.0;
+            for (var skill in questionSkillData) {
+              var totalScorePerSkill = 0.0;
+              var questionIds = questionSkillData[skill].question_ids;
+              var description = questionSkillData[skill].skill_description;
+              for (var i = 0; i < questionIds.length; i += 1) {
+                totalScorePerSkill += questionScores[questionIds[i]];
+              }
+              if (totalScorePerSkill < minScore) {
+                minScore = totalScorePerSkill;
+                ctrl.worstSkill = {
+                  skill_id: skill,
+                  skill_description: description
+                };
+              }
+              $scope.scorePerSkill.push([description, totalScorePerSkill]);
+              totalScore += totalScorePerSkill;
+            }
+            $scope.totalScore = (totalScore * 100) /
+            Object.keys(questionScores).length;
+          };
+
+
+          var calculateScores = function(questionStateData) {
+            $scope.resultsLoaded = false;
+            var questionScores = {};
+            var questionIds = [];
+            for (var question in questionStateData) {
+              questionIds.push(question);
+              var questionData = questionStateData[question];
+              var totalHintsPenalty = 0.0;
+              var wrongAnswerPenalty = 0.0;
+              if (questionData.answers) {
+                wrongAnswerPenalty = (
+                  (questionData.answers.length - 1) * WRONG_ANSWER_PENALTY);
+              }
+              if (questionData.usedHints) {
+                totalHintsPenalty = (
+                  questionData.usedHints.length * VIEW_HINT_PENALTY);
+              }
+              var totalScore = 1.0;
+              if (questionData.viewedSolution) {
+                totalScore = 0.0;
+              } else {
+                totalScore -= (totalHintsPenalty + wrongAnswerPenalty);
+              }
+              questionScores[question] = totalScore;
+            }
+            QuestionPlayerBackendApiService.fetchSkillsForQuestions(
+              questionIds).then(function(result) {
+              calculateScorePerSkill(result, questionScores);
+              $scope.resultsLoaded = true;
+            });
+          };
+
           $rootScope.$on('currentQuestionChanged', function(event, result) {
             updateCurrentQuestion(result + 1);
           });
+
           $rootScope.$on('totalQuestionsReceived', function(event, result) {
             updateTotalQuestions(result);
           });
+
+          $rootScope.$on('questionSessionCompleted', function(event, result) {
```

Comment:
I just wanted to call this part out specifically @seanlip - I thought of adding an encoded version of the question player results as a hash param in the URL, so that logged-out users can also go back and see the results page through their search history (and can even bookmark the page). What do you think?